In [31]:
# les imports 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
from scipy.stats import chi2_contingency
import warnings
warnings.filterwarnings('ignore')

## Sample submission

In [20]:
# Charger le CSV
sample_submission = pd.read_csv("../data/sample_submission.csv")

# Nombre de lignes et colonnes
print("Shape :", sample_submission.shape)

# Aperçu des 5 premières lignes
print(sample_submission.head())

# Doublons
print("Nombre de doublons :", sample_submission.duplicated().sum())

# Valeurs uniques de isFraud
print("Valeurs de isFraud :", sample_submission['isFraud'].unique())

# Nombre d’occurrences de chaque valeur
print(sample_submission['isFraud'].value_counts())

# Statistiques générales
print(sample_submission.describe())

Shape : (506691, 2)
   TransactionID  isFraud
0        3663549      0.5
1        3663550      0.5
2        3663551      0.5
3        3663552      0.5
4        3663553      0.5
Nombre de doublons : 0
Valeurs de isFraud : [0.5]
isFraud
0.5    506691
Name: count, dtype: int64
       TransactionID   isFraud
count   5.066910e+05  506691.0
mean    3.916894e+06       0.5
std     1.462692e+05       0.0
min     3.663549e+06       0.5
25%     3.790222e+06       0.5
50%     3.916894e+06       0.5
75%     4.043566e+06       0.5
max     4.170239e+06       0.5


on a 506691 fraudes dans notre base de données "sample_submission"

## Transaction


In [21]:

###################################  TRAIN  #############################################
print ('=' * 60)
print ('         Train transaction ')
print ('=' * 60)
# Charger les données de transaction
train_transaction = pd.read_csv("../data/train_transaction.csv")

# Shape : nombre de lignes / colonnes
print("Shape :", train_transaction.shape)

# Aperçu
print("\nAperçu :")
display(train_transaction.head())

# Doublons
print("\nNombre de doublons :", train_transaction.duplicated().sum())

# Types des colonnes
print("\nTypes des colonnes :")
print(train_transaction.dtypes)

# Statistiques :
print("\nStatistiques :")
display(train_transaction.describe(include='all'))  # inclut numériques + catégories

# Valeurs manquantes
print("\nNombre de valeurs manquantes par colonne :")
display(train_transaction.isna().sum())

# Colonnes avec le plus de valeurs manquantes
print("\nTop 10 colonnes avec le plus de NaN :")
display(train_transaction.isna().sum().sort_values(ascending=False).head(10))

# Valeurs uniques des colonnes catégorielles
print("\nColonnes catégorielles et nombre de valeurs uniques :")
cat_cols = train_transaction.select_dtypes(include=['object']).columns
for col in cat_cols:
    print(f"{col} : {train_transaction[col].nunique()} valeurs uniques")


######################################################  TEST  #################################################
print ('=' * 60)
print ('         Test transaction ')
print ('=' * 60)
# Charger les données de transaction
test_transaction = pd.read_csv("../data/test_transaction.csv")

# Shape : nombre de lignes / colonnes
print("Shape :", test_transaction.shape)

# Aperçu
print("\nAperçu :")
display(test_transaction.head())

# Doublons
print("\nNombre de doublons :", test_transaction.duplicated().sum())

# Types des colonnes
print("\nTypes des colonnes :")
print(test_transaction.dtypes)

# Statistiques :
print("\nStatistiques :")
display(test_transaction.describe(include='all'))  # inclut numériques + catégories

# Valeurs manquantes
print("\nNombre de valeurs manquantes par colonne :")
display(train_transaction.isna().sum())

# Colonnes avec le plus de valeurs manquantes
print("\nTop 10 colonnes avec le plus de NaN :")
display(train_transaction.isna().sum().sort_values(ascending=False).head(10))

# Valeurs uniques des colonnes catégorielles
print("\nColonnes catégorielles et nombre de valeurs uniques :")
cat_cols = train_transaction.select_dtypes(include=['object']).columns
for col in cat_cols:
    print(f"{col} : {train_transaction[col].nunique()} valeurs uniques")



         Train transaction 
Shape : (590540, 394)

Aperçu :


TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5  ... V330  V331  V332  V333  V334 V335  \
0    NaN  150.0    discover  142.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
1  404.0  150.0  mastercard  102.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
2  490.0  150.0        visa  166.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
3  567.0  150.0  mastercard  117.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
4  514.0  150.0  mastercard  102.0  ...  0.0   0.0   0.0   0.0   0.0  0.0   

  V336  V337  V338  V339  
0  NaN   NaN   NaN   NaN  
1  NaN   NaN   NaN   NaN  
2  NaN   NaN   NaN   NaN  
3  NaN   NaN   NaN   NaN  
4  0.0   0.0   0.0   0.0  

[5 rows x 394 columns]


Nombre de doublons : 0

Types des colonnes :
TransactionID       int64
isFraud             int64
TransactionDT       int64
TransactionAmt    float64
ProductCD          object
                   ...   
V335              float64
V336              float64
V337              float64
V338              float64
V339              float64
Length: 394, dtype: object

Statistiques :


TransactionID        isFraud  TransactionDT  TransactionAmt ProductCD  \
count    5.905400e+05  590540.000000   5.905400e+05   590540.000000    590540   
unique            NaN            NaN            NaN             NaN         5   
top               NaN            NaN            NaN             NaN         W   
freq              NaN            NaN            NaN             NaN    439670   
mean     3.282270e+06       0.034990   7.372311e+06      135.027176       NaN   
std      1.704744e+05       0.183755   4.617224e+06      239.162522       NaN   
min      2.987000e+06       0.000000   8.640000e+04        0.251000       NaN   
25%      3.134635e+06       0.000000   3.027058e+06       43.321000       NaN   
50%      3.282270e+06       0.000000   7.306528e+06       68.769000       NaN   
75%      3.429904e+06       0.000000   1.124662e+07      125.000000       NaN   
max      3.577539e+06       1.000000   1.581113e+07    31937.391000       NaN   

                card1          card2          card3   card4          card5  \
count   590540.000000  581607.000000  588975.000000  588963  586281.000000   
unique            NaN            NaN            NaN       4            NaN   
top               NaN            NaN            NaN    visa            NaN   
freq              NaN            NaN            NaN  384767            NaN   
mean      9898.734658     362.555488     153.194925     NaN     199.278897   
std       4901.170153     157.793246      11.336444     NaN      41.244453   
min       1000.000000     100.000000     100.000000     NaN     100.000000   
25%       6019.000000     214.000000     150.000000     NaN     166.000000   
50%       9678.000000     361.000000     150.000000     NaN     226.000000   
75%      14184.000000     512.000000     150.000000     NaN     226.000000   
max      18396.000000     600.000000     231.000000     NaN     237.000000   

        ...          V330           V331           V332           V333  \
count   ...  82351.000000   82351.000000   82351.000000   82351.000000   
unique  ...           NaN            NaN            NaN            NaN   
top     ...           NaN            NaN            NaN            NaN   
freq    ...           NaN            NaN            NaN            NaN   
mean    ...      0.775874     721.741883    1375.783644    1014.622782   
std     ...      4.727971    6217.223583   11169.275702    7955.735482   
min     ...      0.000000       0.000000       0.000000       0.000000   
25%     ...      0.000000       0.000000       0.000000       0.000000   
50%     ...      0.000000       0.000000       0.000000       0.000000   
75%     ...      0.000000       0.000000      25.000000       0.000000   
max     ...     55.000000  160000.000000  160000.000000  160000.000000   

                V334         V335          V336           V337           V338  \
count   82351.000000  82351.00000  82351.000000   82351.000000   82351.000000   
unique           NaN          NaN           NaN            NaN            NaN   
top              NaN          NaN           NaN            NaN            NaN   
freq             NaN          NaN           NaN            NaN            NaN   
mean        9.807015     59.16455     28.530903      55.352422     151.160542   
std       243.861391    387.62948    274.576920     668.486833    1095.034387   
min         0.000000      0.00000      0.000000       0.000000       0.000000   
25%         0.000000      0.00000      0.000000       0.000000       0.000000   
50%         0.000000      0.00000      0.000000       0.000000       0.000000   
75%         0.000000      0.00000      0.000000       0.000000       0.000000   
max     55125.000000  55125.00000  55125.000000  104060.000000  104060.000000   

                 V339  
count    82351.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean       100.700882  
std        814.946722  
min          0.000000  
25%          0.000000  
50%          0.00000


Nombre de valeurs manquantes par colonne :


TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
V335              508189
V336              508189
V337              508189
V338              508189
V339              508189
Length: 394, dtype: int64


Top 10 colonnes avec le plus de NaN :


dist2    552913
D7       551623
D13      528588
D14      528353
D12      525823
D6       517353
D9       515614
D8       515614
V153     508595
V149     508595
dtype: int64


Colonnes catégorielles et nombre de valeurs uniques :
ProductCD : 5 valeurs uniques
card4 : 4 valeurs uniques
card6 : 4 valeurs uniques
P_emaildomain : 59 valeurs uniques
R_emaildomain : 60 valeurs uniques
M1 : 2 valeurs uniques
M2 : 2 valeurs uniques
M3 : 2 valeurs uniques
M4 : 3 valeurs uniques
M5 : 2 valeurs uniques
M6 : 2 valeurs uniques
M7 : 2 valeurs uniques
M8 : 2 valeurs uniques
M9 : 2 valeurs uniques
         Test transaction 
Shape : (506691, 393)

Aperçu :


TransactionID  TransactionDT  TransactionAmt ProductCD  card1  card2  \
0        3663549       18403224           31.95         W  10409  111.0   
1        3663550       18403263           49.00         W   4272  111.0   
2        3663551       18403310          171.00         W   4476  574.0   
3        3663552       18403310          284.95         W  10989  360.0   
4        3663553       18403317           67.95         W  18018  452.0   

   card3       card4  card5  card6  ...  V330  V331  V332  V333 V334 V335  \
0  150.0        visa  226.0  debit  ...   NaN   NaN   NaN   NaN  NaN  NaN   
1  150.0        visa  226.0  debit  ...   NaN   NaN   NaN   NaN  NaN  NaN   
2  150.0        visa  226.0  debit  ...   NaN   NaN   NaN   NaN  NaN  NaN   
3  150.0        visa  166.0  debit  ...   NaN   NaN   NaN   NaN  NaN  NaN   
4  150.0  mastercard  117.0  debit  ...   NaN   NaN   NaN   NaN  NaN  NaN   

   V336  V337  V338  V339  
0   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN  
3   NaN   NaN   NaN   NaN  
4   NaN   NaN   NaN   NaN  

[5 rows x 393 columns]


Nombre de doublons : 0

Types des colonnes :
TransactionID       int64
TransactionDT       int64
TransactionAmt    float64
ProductCD          object
card1               int64
                   ...   
V335              float64
V336              float64
V337              float64
V338              float64
V339              float64
Length: 393, dtype: object

Statistiques :


TransactionID  TransactionDT  TransactionAmt ProductCD          card1  \
count    5.066910e+05   5.066910e+05   506691.000000    506691  506691.000000   
unique            NaN            NaN             NaN         5            NaN   
top               NaN            NaN             NaN         W            NaN   
freq              NaN            NaN             NaN    360987            NaN   
mean     3.916894e+06   2.692994e+07      134.725568       NaN    9957.222175   
std      1.462692e+05   4.756507e+06      245.779822       NaN    4884.960969   
min      3.663549e+06   1.840322e+07        0.018000       NaN    1001.000000   
25%      3.790222e+06   2.277154e+07       40.000000       NaN    6019.000000   
50%      3.916894e+06   2.720466e+07       67.950000       NaN    9803.000000   
75%      4.043566e+06   3.134856e+07      125.000000       NaN   14276.000000   
max      4.170239e+06   3.421434e+07    10270.000000       NaN   18397.000000   

                card2          card3   card4          card5   card6  ...  \
count   498037.000000  503689.000000  503605  502144.000000  503684  ...   
unique            NaN            NaN       4            NaN       3  ...   
top               NaN            NaN    visa            NaN   debit  ...   
freq              NaN            NaN  334882            NaN  385021  ...   
mean       363.735379     153.543409     NaN     200.162975     NaN  ...   
std        158.688653      12.443013     NaN      40.562461     NaN  ...   
min        100.000000     100.000000     NaN     100.000000     NaN  ...   
25%        207.000000     150.000000     NaN     166.000000     NaN  ...   
50%        369.000000     150.000000     NaN     226.000000     NaN  ...   
75%        512.000000     150.000000     NaN     226.000000     NaN  ...   
max        600.000000     232.000000     NaN     237.000000     NaN  ...   

                V330          V331          V332          V333          V334  \
count   76431.000000  7.643100e+04  7.643100e+04  7.643100e+04  76431.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.309900  3.151286e+02  4.404643e+02  3.811021e+02     24.708030   
std         2.632181  9.494045e+03  1.064843e+04  1.039556e+04    842.913287   
min         0.000000  0.000000e+00  0.000000e+00  0.000000e+00      0.000000   
25%         0.000000  0.000000e+00  0.000000e+00  0.000000e+00      0.000000   
50%         0.000000  0.000000e+00  0.000000e+00  0.000000e+00      0.000000   
75%         0.000000  0.000000e+00  0.000000e+00  0.000000e+00      0.000000   
max       106.000000  1.040658e+06  1.040658e+06  1.040658e+06  64800.000000   

                V335          V336           V337           V338  \
count   76431.000000  76431.000000   76431.000000   76431.000000   
unique           NaN           NaN            NaN            NaN   
top              NaN           NaN            NaN            NaN   
freq             NaN           NaN            NaN            NaN   
mean       58.339468     35.485643      99.058760     155.578153   
std       990.361078    904.093921    2436.074522    4544.303974   
min         0.000000      0.000000       0.000000       0.000000   
25%         0.000000      0.000000       0.000000       0.000000   
50%         0.000000      0.000000       0.000000       0.000000   
75%         0.000000      0.000000       0.000000       0.000000   
max     64800.000000  64800.000000  375000.000000  612500.000000   

                 V339  
count    76431.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean       139.814706  
std       4493.728267  
min          0.000000  
25%          0.000000  
50%          0.000000  
75%          0.000000  
max     612500.000000  

[11 rows x 393 columns]


Nombre de valeurs manquantes par colonne :


TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
V335              508189
V336              508189
V337              508189
V338              508189
V339              508189
Length: 394, dtype: int64


Top 10 colonnes avec le plus de NaN :


dist2    552913
D7       551623
D13      528588
D14      528353
D12      525823
D6       517353
D9       515614
D8       515614
V153     508595
V149     508595
dtype: int64


Colonnes catégorielles et nombre de valeurs uniques :
ProductCD : 5 valeurs uniques
card4 : 4 valeurs uniques
card6 : 4 valeurs uniques
P_emaildomain : 59 valeurs uniques
R_emaildomain : 60 valeurs uniques
M1 : 2 valeurs uniques
M2 : 2 valeurs uniques
M3 : 2 valeurs uniques
M4 : 3 valeurs uniques
M5 : 2 valeurs uniques
M6 : 2 valeurs uniques
M7 : 2 valeurs uniques
M8 : 2 valeurs uniques
M9 : 2 valeurs uniques


Il y a 

## Identity

In [22]:
#####################################  TRAIN  ############################################################
print ('=' * 60)
print ('         Train Identity ')
print ('=' * 60)
# Charger les données identity
train_identity = pd.read_csv("../data/train_identity.csv")

# Shape : nombre de lignes / colonnes
print("Shape :", train_identity.shape)

# Aperçu
print("\nAperçu :")
display(train_identity.head())

# Doublons
print("\nNombre de doublons :", train_identity.duplicated().sum())

# Types des colonnes
print("\nTypes des colonnes :")
print(train_identity.dtypes)

# Statistiques :
print("\nStatistiques :")
display(train_identity.describe(include='all'))  # inclut numériques + catégories

# Valeurs manquantes
print("\nNombre de valeurs manquantes par colonne :")
display(train_identity.isna().sum())

# Colonnes avec le plus de valeurs manquantes
print("\nTop 10 colonnes avec le plus de NaN :")
display(train_identity.isna().sum().sort_values(ascending=False).head(10))

# Valeurs uniques des colonnes catégorielles
print("\nColonnes catégorielles et nombre de valeurs uniques :")
cat_cols = train_identity.select_dtypes(include=['object']).columns
for col in cat_cols:
    print(f"{col} : {train_identity[col].nunique()} valeurs uniques")

#####################################  TEST ############################################################
print ('=' * 60)
print ('         Test Identity ')
print ('=' * 60)

# Charger les données identity
test_identity = pd.read_csv("../data/test_identity.csv")

# Shape : nombre de lignes / colonnes
print("Shape :", test_identity.shape)

# Aperçu
print("\nAperçu :")
display(test_identity.head())

# Doublons
print("\nNombre de doublons :", test_identity.duplicated().sum())

# Types des colonnes
print("\nTypes des colonnes :")
print(test_identity.dtypes)

# Statistiques :
print("\nStatistiques :")
display(test_identity.describe(include='all'))  # inclut numériques + catégories

# Valeurs manquantes
print("\nNombre de valeurs manquantes par colonne :")
display(test_identity.isna().sum())

# Colonnes avec le plus de valeurs manquantes
print("\nTop 10 colonnes avec le plus de NaN :")
display(test_identity.isna().sum().sort_values(ascending=False).head(10))

# Valeurs uniques des colonnes catégorielles
print("\nColonnes catégorielles et nombre de valeurs uniques :")
cat_cols = test_identity.select_dtypes(include=['object']).columns
for col in cat_cols:
    print(f"{col} : {test_identity[col].nunique()} valeurs uniques")

         Train Identity 
Shape : (144233, 41)

Aperçu :


TransactionID  id_01     id_02  id_03  id_04  id_05  id_06  id_07  id_08  \
0        2987004    0.0   70787.0    NaN    NaN    NaN    NaN    NaN    NaN   
1        2987008   -5.0   98945.0    NaN    NaN    0.0   -5.0    NaN    NaN   
2        2987010   -5.0  191631.0    0.0    0.0    0.0    0.0    NaN    NaN   
3        2987011   -5.0  221832.0    NaN    NaN    0.0   -6.0    NaN    NaN   
4        2987016    0.0    7460.0    0.0    0.0    1.0    0.0    NaN    NaN   

   id_09  ...                id_31  id_32      id_33           id_34  id_35  \
0    NaN  ...  samsung browser 6.2   32.0  2220x1080  match_status:2      T   
1    NaN  ...   mobile safari 11.0   32.0   1334x750  match_status:1      T   
2    0.0  ...          chrome 62.0    NaN        NaN             NaN      F   
3    NaN  ...          chrome 62.0    NaN        NaN             NaN      F   
4    0.0  ...          chrome 62.0   24.0   1280x800  match_status:2      T   

  id_36 id_37  id_38  DeviceType                     DeviceInfo  
0     F     T      T      mobile  SAMSUNG SM-G892A Build/NRD90M  
1     F     F      T      mobile                     iOS Device  
2     F     T      T     desktop                        Windows  
3     F     T      T     desktop                            NaN  
4     F     T      T     desktop                          MacOS  

[5 rows x 41 columns]


Nombre de doublons : 0

Types des colonnes :
TransactionID      int64
id_01            float64
id_02            float64
id_03            float64
id_04            float64
id_05            float64
id_06            float64
id_07            float64
id_08            float64
id_09            float64
id_10            float64
id_11            float64
id_12             object
id_13            float64
id_14            float64
id_15             object
id_16             object
id_17            float64
id_18            float64
id_19            float64
id_20            float64
id_21            float64
id_22            float64
id_23             object
id_24            float64
id_25            float64
id_26            float64
id_27             object
id_28             object
id_29             object
id_30             object
id_31             object
id_32            float64
id_33             object
id_34             object
id_35             object
id_36             object
id_37             object
id_3

TransactionID          id_01          id_02         id_03  \
count    1.442330e+05  144233.000000  140872.000000  66324.000000   
unique            NaN            NaN            NaN           NaN   
top               NaN            NaN            NaN           NaN   
freq              NaN            NaN            NaN           NaN   
mean     3.236329e+06     -10.170502  174716.584708      0.060189   
std      1.788496e+05      14.347949  159651.816856      0.598231   
min      2.987004e+06    -100.000000       1.000000    -13.000000   
25%      3.077142e+06     -10.000000   67992.000000      0.000000   
50%      3.198818e+06      -5.000000  125800.500000      0.000000   
75%      3.392923e+06      -5.000000  228749.000000      0.000000   
max      3.577534e+06       0.000000  999595.000000     10.000000   

               id_04          id_05          id_06        id_07        id_08  \
count   66324.000000  136865.000000  136865.000000  5155.000000  5155.000000   
unique           NaN            NaN            NaN          NaN          NaN   
top              NaN            NaN            NaN          NaN          NaN   
freq             NaN            NaN            NaN          NaN          NaN   
mean       -0.058938       1.615585      -6.698710    13.285354   -38.600388   
std         0.701015       5.249856      16.491104    11.384207    26.084899   
min       -28.000000     -72.000000    -100.000000   -46.000000  -100.000000   
25%         0.000000       0.000000      -6.000000     5.000000   -48.000000   
50%         0.000000       0.000000       0.000000    14.000000   -34.000000   
75%         0.000000       1.000000       0.000000    22.000000   -23.000000   
max         0.000000      52.000000       0.000000    61.000000     0.000000   

               id_09  ...        id_31         id_32      id_33  \
count   74926.000000  ...       140282  77586.000000      73289   
unique           NaN  ...          130           NaN        260   
top              NaN  ...  chrome 63.0           NaN  1920x1080   
freq             NaN  ...        22000           NaN      16874   
mean        0.091023  ...          NaN     26.508597        NaN   
std         0.983842  ...          NaN      3.737502        NaN   
min       -36.000000  ...          NaN      0.000000        NaN   
25%         0.000000  ...          NaN     24.000000        NaN   
50%         0.000000  ...          NaN     24.000000        NaN   
75%         0.000000  ...          NaN     32.000000        NaN   
max        25.000000  ...          NaN     32.000000        NaN   

                 id_34   id_35   id_36   id_37   id_38  DeviceType  DeviceInfo  
count            77805  140985  140985  140985  140985      140810      118666  
unique               4       2       2       2       2           2        1786  
top     match_status:2       T       F       T       F     desktop     Windows  
freq             60011   77814  134066  110452   73922       85165       47722  
mean               NaN     NaN     NaN     NaN     NaN         NaN         NaN  
std                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
min                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
25%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
50%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
75%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
max                NaN     NaN     NaN     NaN     NaN         NaN         NaN  

[11 rows x 41 columns]


Nombre de valeurs manquantes par colonne :


TransactionID         0
id_01                 0
id_02              3361
id_03             77909
id_04             77909
id_05              7368
id_06              7368
id_07            139078
id_08            139078
id_09             69307
id_10             69307
id_11              3255
id_12                 0
id_13             16913
id_14             64189
id_15              3248
id_16             14893
id_17              4864
id_18             99120
id_19              4915
id_20              4972
id_21            139074
id_22            139064
id_23            139064
id_24            139486
id_25            139101
id_26            139070
id_27            139064
id_28              3255
id_29              3255
id_30             66668
id_31              3951
id_32             66647
id_33             70944
id_34             66428
id_35              3248
id_36              3248
id_37              3248
id_38              3248
DeviceType         3423
DeviceInfo        25567
dtype: int64


Top 10 colonnes avec le plus de NaN :


id_24    139486
id_25    139101
id_07    139078
id_08    139078
id_21    139074
id_26    139070
id_23    139064
id_27    139064
id_22    139064
id_18     99120
dtype: int64


Colonnes catégorielles et nombre de valeurs uniques :
id_12 : 2 valeurs uniques
id_15 : 3 valeurs uniques
id_16 : 2 valeurs uniques
id_23 : 3 valeurs uniques
id_27 : 2 valeurs uniques
id_28 : 2 valeurs uniques
id_29 : 2 valeurs uniques
id_30 : 75 valeurs uniques
id_31 : 130 valeurs uniques
id_33 : 260 valeurs uniques
id_34 : 4 valeurs uniques
id_35 : 2 valeurs uniques
id_36 : 2 valeurs uniques
id_37 : 2 valeurs uniques
id_38 : 2 valeurs uniques
DeviceType : 2 valeurs uniques
DeviceInfo : 1786 valeurs uniques
         Test Identity 
Shape : (141907, 41)

Aperçu :


TransactionID  id-01     id-02  id-03  id-04  id-05  id-06  id-07  id-08  \
0        3663586  -45.0  280290.0    NaN    NaN    0.0    0.0    NaN    NaN   
1        3663588    0.0    3579.0    0.0    0.0    0.0    0.0    NaN    NaN   
2        3663597   -5.0  185210.0    NaN    NaN    1.0    0.0    NaN    NaN   
3        3663601  -45.0  252944.0    0.0    0.0    0.0    0.0    NaN    NaN   
4        3663602  -95.0  328680.0    NaN    NaN    7.0  -33.0    NaN    NaN   

   id-09  ...                    id-31  id-32     id-33           id-34  \
0    NaN  ...  chrome 67.0 for android    NaN       NaN             NaN   
1    0.0  ...  chrome 67.0 for android   24.0  1280x720  match_status:2   
2    NaN  ...       ie 11.0 for tablet    NaN       NaN             NaN   
3    0.0  ...  chrome 67.0 for android    NaN       NaN             NaN   
4    NaN  ...  chrome 67.0 for android    NaN       NaN             NaN   

   id-35 id-36 id-37  id-38  DeviceType                   DeviceInfo  
0      F     F     T      F      mobile  MYA-L13 Build/HUAWEIMYA-L13  
1      T     F     T      T      mobile         LGLS676 Build/MXB48T  
2      F     T     T      F     desktop                  Trident/7.0  
3      F     F     T      F      mobile  MYA-L13 Build/HUAWEIMYA-L13  
4      F     F     T      F      mobile         SM-G9650 Build/R16NW  

[5 rows x 41 columns]


Nombre de doublons : 0

Types des colonnes :
TransactionID      int64
id-01            float64
id-02            float64
id-03            float64
id-04            float64
id-05            float64
id-06            float64
id-07            float64
id-08            float64
id-09            float64
id-10            float64
id-11            float64
id-12             object
id-13            float64
id-14            float64
id-15             object
id-16             object
id-17            float64
id-18            float64
id-19            float64
id-20            float64
id-21            float64
id-22            float64
id-23             object
id-24            float64
id-25            float64
id-26            float64
id-27             object
id-28             object
id-29             object
id-30             object
id-31             object
id-32            float64
id-33             object
id-34             object
id-35             object
id-36             object
id-37             object
id-3

TransactionID          id-01          id-02         id-03  \
count    1.419070e+05  141907.000000  136976.000000  66481.000000   
unique            NaN            NaN            NaN           NaN   
top               NaN            NaN            NaN           NaN   
freq              NaN            NaN            NaN           NaN   
mean     3.972166e+06     -11.325734  192658.729909      0.053008   
std      1.469966e+05      14.508520  182613.277215      0.684551   
min      3.663586e+06    -100.000000       2.000000    -12.000000   
25%      3.859268e+06     -12.500000   63339.500000      0.000000   
50%      4.001774e+06      -5.000000  133189.500000      0.000000   
75%      4.105284e+06      -5.000000  265717.500000      0.000000   
max      4.170239e+06       0.000000  999869.000000     11.000000   

               id-04          id-05          id-06        id-07        id-08  \
count   66481.000000  134750.000000  134750.000000  5059.000000  5059.000000   
unique           NaN            NaN            NaN          NaN          NaN   
top              NaN            NaN            NaN          NaN          NaN   
freq             NaN            NaN            NaN          NaN          NaN   
mean       -0.087454       1.246033      -6.803829    12.493180   -36.577782   
std         0.840351       5.071394      15.921457    11.678206    25.544185   
min       -19.000000     -81.000000    -100.000000   -41.000000  -100.000000   
25%         0.000000       0.000000      -6.000000     3.000000   -46.000000   
50%         0.000000       0.000000       0.000000    12.000000   -33.000000   
75%         0.000000       1.000000       0.000000    21.000000   -23.000000   
max         0.000000      52.000000       0.000000    59.000000     0.000000   

               id-09  ...        id-31         id-32      id-33  \
count   74338.000000  ...       136625  70671.000000      70671   
unique           NaN  ...          135           NaN        390   
top              NaN  ...  chrome 70.0           NaN  1920x1080   
freq             NaN  ...        16054           NaN      16868   
mean        0.076219  ...          NaN     26.217939        NaN   
std         1.009687  ...          NaN      3.601046        NaN   
min       -32.000000  ...          NaN      8.000000        NaN   
25%         0.000000  ...          NaN     24.000000        NaN   
50%         0.000000  ...          NaN     24.000000        NaN   
75%         0.000000  ...          NaN     32.000000        NaN   
max        16.000000  ...          NaN     48.000000        NaN   

                 id-34   id-35   id-36   id-37   id-38  DeviceType  DeviceInfo  
count            72175  136977  136977  136977  136977      136931      115057  
unique               2       2       2       2       2           2        2226  
top     match_status:2       T       F       T       F     desktop     Windows  
freq             72174   71650  133287  104697   95058       74403       44988  
mean               NaN     NaN     NaN     NaN     NaN         NaN         NaN  
std                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
min                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
25%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
50%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
75%                NaN     NaN     NaN     NaN     NaN         NaN         NaN  
max                NaN     NaN     NaN     NaN     NaN         NaN         NaN  

[11 rows x 41 columns]


Nombre de valeurs manquantes par colonne :


TransactionID         0
id-01                 0
id-02              4931
id-03             75426
id-04             75426
id-05              7157
id-06              7157
id-07            136848
id-08            136848
id-09             67569
id-10             67569
id-11              5129
id-12                 0
id-13             11621
id-14             70550
id-15              4930
id-16             16160
id-17              5941
id-18             91032
id-19              6001
id-20              6274
id-21            136848
id-22            136845
id-23            136845
id-24            137167
id-25            136868
id-26            136860
id-27            136845
id-28              5129
id-29              5129
id-30             71248
id-31              5282
id-32             71236
id-33             71236
id-34             69732
id-35              4930
id-36              4930
id-37              4930
id-38              4930
DeviceType         4976
DeviceInfo        26850
dtype: int64


Top 10 colonnes avec le plus de NaN :


id-24    137167
id-25    136868
id-26    136860
id-08    136848
id-07    136848
id-21    136848
id-23    136845
id-27    136845
id-22    136845
id-18     91032
dtype: int64


Colonnes catégorielles et nombre de valeurs uniques :
id-12 : 2 valeurs uniques
id-15 : 3 valeurs uniques
id-16 : 2 valeurs uniques
id-23 : 3 valeurs uniques
id-27 : 2 valeurs uniques
id-28 : 2 valeurs uniques
id-29 : 2 valeurs uniques
id-30 : 86 valeurs uniques
id-31 : 135 valeurs uniques
id-33 : 390 valeurs uniques
id-34 : 2 valeurs uniques
id-35 : 2 valeurs uniques
id-36 : 2 valeurs uniques
id-37 : 2 valeurs uniques
id-38 : 2 valeurs uniques
DeviceType : 2 valeurs uniques
DeviceInfo : 2226 valeurs uniques


Il y a 144 233 identités dans la base de données identité

# Nouvelle base de données de simulations de transactions 

In [39]:
print('='*60)
print('          Train Fraud')
print('='*60)
train_fraud = pd.read_csv("../data/SimulatedfraudTrain.csv")
print(f"Nombre de transactions: {len(train_fraud):,}")
print(f"Nombre de variables: {train_fraud.shape[1]}")
print(train_fraud.describe())

# valeurs manquantes
print ('     valeurs manquantes ')
missing = train_fraud.isnull().sum()
missing_pct = (missing / len(train_fraud)) * 100
missing_df = pd.DataFrame({
    'Colonne': missing.index,
    'Valeurs_Manquantes': missing.values,
    'Pourcentage': missing_pct.values
})
missing_df = missing_df[missing_df['Valeurs_Manquantes'] > 0].sort_values('Valeurs_Manquantes', ascending=False)
if len(missing_df) > 0:
    print("Colonnes avec valeurs manquantes:")
    print(missing_df.to_string(index=False))
    
    # Visualisation
    if len(missing_df) > 0:
        plt.figure(figsize=(12, 6))
        plt.barh(missing_df['Colonne'], missing_df['Pourcentage'], color='coral')
        plt.xlabel('Pourcentage de valeurs manquantes')
        plt.title('Valeurs Manquantes par Variable', fontsize=14, fontweight='bold')
        plt.tight_layout()
        plt.show()
else:
    print("Aucune valeur manquante détectée!")


# Fraud 

# Compter les valeurs 0 et 1
print("Nombre de transactions par type:")
print(train_fraud['is_fraud'].value_counts())

print("\nPourcentages:")
print(train_fraud['is_fraud'].value_counts(normalize=True) * 100)

# Ou de façon plus détaillée
nb_non_fraude = (train_fraud['is_fraud'] == 0).sum()
nb_fraude = (train_fraud['is_fraud'] == 1).sum()

print(f"\n0 (Non-Fraude): {nb_non_fraude:,}")
print(f"1 (Fraude): {nb_fraude:,}")
print(f"Total: {len(train_fraud):,}")

# gender 

print("Nombre de transactions pour chaque type de genre")
print(train_fraud['gender'].value_counts())

print("\nPourcentages:")
print(train_fraud['gender'].value_counts(normalize=True) * 100)

# Ou de façon plus détaillée
nb_masculin = (train_fraud['gender'] == 'M').sum()
nb_feminin = (train_fraud['gender'] == 'F').sum()

print(f"\nM (Masculin): {nb_masculin:,}")
print(f"F (Féminin): {nb_feminin:,}")
print(f"Total: {len(train_fraud):,}")


# Vérification de l'uniicité de la colonne "cc_num"

# Nombre total de lignes
n_total = len(train_fraud)

# Nombre de valeurs uniques de cc_num
n_unique = train_fraud['cc_num'].nunique()

print("Nombre total de transactions :", n_total)
print("Nombre de cc_num uniques     :", n_unique)

if n_unique == n_total:
    print("La colonne cc_num est unique (une valeur par transaction).")
else:
    print("La colonne cc_num n'est PAS unique (une carte a plusieurs transactions).")


# vérification de l'unicité de la colonne 'trans_num'
# Nombre de valeurs uniques de trans_num
n_unique = train_fraud['trans_num'].nunique()

print("Nombre total de transactions :", n_total)
print("Nombre de trans_num uniques     :", n_unique)

if n_unique == n_total:
    print("La colonne trans_num est unique (une valeur par transaction).")
else:
    print("La colonne trans_num n'est PAS unique .")

          Train Fraud
Nombre de transactions: 1,296,675
Nombre de variables: 23
         Unnamed: 0        cc_num           amt           zip           lat  \
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06   
mean   6.483370e+05  4.171920e+17  7.035104e+01  4.880067e+04  3.853762e+01   
std    3.743180e+05  1.308806e+18  1.603160e+02  2.689322e+04  5.075808e+00   
min    0.000000e+00  6.041621e+10  1.000000e+00  1.257000e+03  2.002710e+01   
25%    3.241685e+05  1.800429e+14  9.650000e+00  2.623700e+04  3.462050e+01   
50%    6.483370e+05  3.521417e+15  4.752000e+01  4.817400e+04  3.935430e+01   
75%    9.725055e+05  4.642255e+15  8.314000e+01  7.204200e+04  4.194040e+01   
max    1.296674e+06  4.992346e+18  2.894890e+04  9.978300e+04  6.669330e+01   

               long      city_pop     unix_time     merch_lat    merch_long  \
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06   
mean  -9.022634e+01  8.882444e+04  1.349244e+09  3

In [37]:
print('='*60)
print('         Test Fraud')
print('='*60)
test_fraud = pd.read_csv("../data/SimulatedfraudTest.csv")
print(f"Nombre de transactions: {len(test_fraud):,}")
print(f"Nombre de variables: {test_fraud.shape[1]}")
print(test_fraud.describe())

# valeurs manquantes
print ('     valeurs manquantes ')
missing = test_fraud.isnull().sum()
missing_pct = (missing / len(test_fraud)) * 100
missing_df = pd.DataFrame({
    'Colonne': missing.index,
    'Valeurs_Manquantes': missing.values,
    'Pourcentage': missing_pct.values
})
missing_df = missing_df[missing_df['Valeurs_Manquantes'] > 0].sort_values('Valeurs_Manquantes', ascending=False)
if len(missing_df) > 0:
    print("Colonnes avec valeurs manquantes:")
    print(missing_df.to_string(index=False))
    
    # Visualisation
    if len(missing_df) > 0:
        plt.figure(figsize=(12, 6))
        plt.barh(missing_df['Colonne'], missing_df['Pourcentage'], color='coral')
        plt.xlabel('Pourcentage de valeurs manquantes')
        plt.title('Valeurs Manquantes par Variable', fontsize=14, fontweight='bold')
        plt.tight_layout()
        plt.show()
else:
    print("Aucune valeur manquante détectée!")


# Fraud 

# Compter les valeurs 0 et 1
print("Nombre de transactions par type:")
print(test_fraud['is_fraud'].value_counts())

print("\nPourcentages:")
print(test_fraud['is_fraud'].value_counts(normalize=True) * 100)

# Ou de façon plus détaillée
nb_non_fraude = (test_fraud['is_fraud'] == 0).sum()
nb_fraude = (test_fraud['is_fraud'] == 1).sum()

print(f"\n0 (Non-Fraude): {nb_non_fraude:,}")
print(f"1 (Fraude): {nb_fraude:,}")
print(f"Total: {len(train_fraud):,}")

# gender 

print("Nombre de transactions pour chaque type de genre")
print(test_fraud['gender'].value_counts())

print("\nPourcentages:")
print(test_fraud['gender'].value_counts(normalize=True) * 100)

# Ou de façon plus détaillée
nb_masculin = (test_fraud['gender'] == 'M').sum()
nb_feminin = (test_fraud['gender'] == 'F').sum()

print(f"\nM (Masculin): {nb_masculin:,}")
print(f"F (Féminin): {nb_feminin:,}")
print(f"Total: {len(test_fraud):,}")


# Vérification de l'uniicité de la colonne "cc_num"
# Nombre total de lignes
n_total = len(test_fraud)

# Nombre de valeurs uniques de cc_num
n_unique = test_fraud['cc_num'].nunique()

print("Nombre total de transactions :", n_total)
print("Nombre de cc_num uniques     :", n_unique)

if n_unique == n_total:
    print("La colonne cc_num est unique (une valeur par transaction).")
else:
    print("La colonne cc_num n'est PAS unique (une carte a plusieurs transactions).")


         Test Fraud
Nombre de transactions: 555,719
Nombre de variables: 23
          Unnamed: 0        cc_num            amt            zip  \
count  555719.000000  5.557190e+05  555719.000000  555719.000000   
mean   277859.000000  4.178387e+17      69.392810   48842.628015   
std    160422.401459  1.309837e+18     156.745941   26855.283328   
min         0.000000  6.041621e+10       1.000000    1257.000000   
25%    138929.500000  1.800429e+14       9.630000   26292.000000   
50%    277859.000000  3.521417e+15      47.290000   48174.000000   
75%    416788.500000  4.635331e+15      83.010000   72011.000000   
max    555718.000000  4.992346e+18   22768.110000   99921.000000   

                 lat           long      city_pop     unix_time  \
count  555719.000000  555719.000000  5.557190e+05  5.557190e+05   
mean       38.543253     -90.231325  8.822189e+04  1.380679e+09   
std         5.061336      13.721780  3.003909e+05  5.201104e+06   
min        20.027100    -165.672300  2.3000

In [28]:
import sys
print(sys.executable)


/home/ensai/Documents/MSD/Smart_data_project/Anomalies-Fraud-Detection/Variational-graph-autoencoder-for-fraud-detection/.venv/bin/python
